### PR TITLE
E2E tests migration to Playwright

### PR DESCRIPTION
**Pull request by adimoldovan**
This is a POC for a migration of E2E tests from Puppeteer to Playwright.

##### Added dependencies:
- playwright
- jest-playwright-preset
- playwright-video
- @ffmpeg-installer/ffmpeg

##### Removed dependencies:
- puppeteer
- jest-puppeteer

##### Configuration changes
- Removed `jest-puppeteer.config.js` and added `jest-playwright.config.js` instead 
- Added `JEST_PLAYWRIGHT_CONFIG` env variable in `jest.config.js` instead of CLI as it never changes
- added `testRunner: 'jasmine2'` in `jest.config.js` to fix `"referenceError: jasmine is not defined"`
- I had an error reading the config files and I could only fix it by changing the format from .js to .json. So `default.js` and `local-tests.js` are now `default.json` and `local-tests.json`

##### Code changes due to API differences
- replaced `waitAndClick` and `waitAndType` with Playwright's `page.click` and `page.type`. Removed `waitAndClick` and `waitAndType` as they are now obsolete
- replaced `page.waitFor` with `page.waitForTimeout`
- replaced `page.setCookie()` with `page.browserContext().addCookies()`
- replaced `page.$x` with `page.$$(xpath='locator')`
- replaced `page.waitForXPath()` with `page.waitForSelector()`
- replaced visibility options `visible: true` and `hidden: true` with `state: 'visible'` and `state: 'hidden'`
- replaced `page.setViewport()` with `page.setViewportSize()`
- replaced `networkidle0` and `networkidle0` with `networkidle`
- updated env variables names, like `PUPPETEER_HEADLES` to `HEADLESS`

##### New features
- Added video capture support, under a toggle controlled by `CAPTURE_VIDEO` env variable. Run with `CAPTURE_VIDEO=true yarn test-e2e` to capture video.

##### Changes unrelated to migration
- Grouped test output (logs, videos, screenshots) under a single directory `tests/e2e/output`. I was not able to move allure-results though
- Added command to clean the output dir before tests run
- Other small tweaks like locators changes to make tests pass or more resilient

## Left to do

- Change user agent
- Make all tests pass. I'm assuming that current failures are not related to the migration, but still, a confirmation is needed.
- There is a lot of room for code cleanup. E.g. some logging may be obsolete now as Playwright has really good debugging logging with `DEBUG=pw:api` option. Possibly some methods in page-helper are also obsolete.

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
All E2E tests need to pass

#### Proposed changelog entry for your changes:
no changelog entry needed

### CODE REVIEW

File: tests/e2e/lib/global-teardown.js [hunk 1]
```diff
@@ -61,5 +60,5 @@ module.exports = async function ( globalConfig ) {
 	if ( process.env.CI ) {
 		await processSlackLog();
 	}
-	await teardown( globalConfig );
+	// await teardown( globalConfig );
```
Comment (kraftbj) [inline tests/e2e/lib/global-teardown.js:63]: Minor nitpick: Is this left here for anything in particular or missed in the cleanup?

File: tests/e2e/README.md [hunk 2]
```diff
@@ -53,10 +53,10 @@ You can run the e2e tests locally using this command:
 yarn test-e2e
 ```
 
-Puppeteer runs headless by default (i.e. browser is not visible). However, sometimes it's useful to observe the browser while running tests. To see the browser window and the running tests you can pass `PUPPETEER_HEADLESS=false` as follows:
+Puppeteer runs headless by default (i.e. browser is not visible). However, sometimes it's useful to observe the browser while running tests. To see the browser window and the running tests you can pass `HEADLESS=false` as follows:
```
Comment (kraftbj) [inline tests/e2e/README.md:56]: There are various references to Puppeteer still in the codebase and documentation.

File: tests/e2e/lib/global-teardown.js (hunk 1 above)
Comment (adimoldovan) [reply tests/e2e/lib/global-teardown.js:64]: Good catch. I actually forgot about this one, tests not complaining about it... 
This was a call to a jest-puppeteer function and I was not sure what to replace it with, or now, even if it should be replaced at all. 
Maybe @brbrr can help.

File: tests/e2e/README.md (hunk 2 above)
Comment (adimoldovan) [reply tests/e2e/README.md]: Indeed, I left a lot of those in documentation blocks mostly. I pushed a change which replaced all except one for a piece o code that still needs to be evaluated if it should stay or go.

File: tests/e2e/lib/global-teardown.js (hunk 1 above)
Comment (brbrr) [reply tests/e2e/lib/global-teardown.js:64]: it coming from `jest-puppeteer` dependency that we using: https://github.com/smooth-code/jest-puppeteer#create-your-own-globalsetup-and-globalteardown

I guess we playwright counterpart does not need that (I'm not sure why tho)

File: tests/e2e/jest.config.js [hunk 3]
```diff
@@ -3,12 +3,15 @@
...
-	preset: 'jest-puppeteer',
+	preset: 'jest-playwright-preset',
 	globalTeardown: './lib/global-teardown.js',
 	setupFilesAfterEnv: [
 		'jest-allure/dist/setup',
 		'<rootDir>/lib/setup-env.js',
-		'expect-puppeteer',
+		'expect-playwright',
 	],
+	testRunner: 'jasmine2',
```
Comment (brbrr) [inline tests/e2e/jest.config.js:16]: I understand that this line was added basically to make tests work. But I wonder if this is an optimal solution.

From what I gathered, with playwright we have a few options for the test runner:

- jasmine2 (which is used in the current configuration)
- jest-circus
- playwright-test

Could you talk about the pros and cons of these as well? I know that `circus` supports test re-run which I find really appealing. Also I wonder what benefits `playwright-test` could provide.

Comment (adimoldovan) [reply tests/e2e/jest.config.js:16]: Yes, this was a fix for an error thrown by jest-allure library. I don't really understand how come it wasn't failing before, with the Puppeteer suite of dependencies.
This was the only working solution I could find, based on this: https://github.com/zaqqaz/jest-allure/issues/8
It's not ideal, it was the way to move forward while keeping the Allure reports.
I will look into other options.

Comment (brbrr) [reply tests/e2e/jest.config.js:16]: Oh, I got it. I thought the source of this error was our own usage of `jasmine`. I guess we would need to move away from `jest-allure` if we'd like to switch to `circus` runner.

I wonder if there are alternatives to `jest-allure` that would work with `circus` runner.

**Comment by ebinnion** [comment]
Leaving a note that we closed this PR in favor of #17814 since this PR was created from a forked repo.

**Pull request closed without merging**